pennylane/ops/pulse/convenience_functions.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This file contains convenience functions for pulse programming."""
+
+import numpy as np
+
+has_jax = True
+try:
+    import jax.numpy as jnp
+except ImportError:
+    has_jax = False
+
+
+def pwc(timespan):
+    """Creates a function that is piecewise-constant in time.
+
+    Args:
+            timespan(Union[float, tuple(float, float)]: The timespan defining the region where the function is non-zero.
+              If an integer is provided, the timespan is defined as ``(0, timespan)``.
+
+    Returns:
+            func: a function that takes two arguments, an array of trainable parameters and a `float` defining the
+            time at which the function is evaluated. When called, the function uses the array of parameters to
+            create evenly sized bins within the ``timespan``, with each bin value set by an element of the array.
+            It then selects the value the parameter array corresponding to the specified time.
+
+    **Example**
+
+    >>> timespan = (1, 3)
+    >>> f1 = pwc(timespan)
+
+    The resulting function ``f1`` has the call signature ``f1(params, t)``. If passed an array of parameters and
+    a time, it will assign the array as the constants in the piecewise function, and select the constant corresponding
+    to the specified time, based on the time interval defined by ``timespan``.
+
+    >>> params = [np.linspace(10, 20, 10)]
+    >>> f1(params, 2)
```

Comment:
Careful! This raises an error:

```
ValueError: setting an array element with a sequence.
```

This is because in the following line you add 0 to the list of params, which raises the error when casting this to a tensor:
```python
# include 0 as an additional option for function output
params = jnp.array(list(params) + [0])

# get idx from timestamp, then set idx=0 if idx is out of bounds for the array
idx = num_bins / (t2 - t1) * (t - t1)
```

Reply:
Should we restrict the `params` input to a flat list? (thus use `np.linspace(10, 20, 10)` instead of `[np.linspace(10, 20, 10)]`)

Reply:
Yes, thanks for catching that! The nested list is a relic from when the example passed `params` to a `ParametrizedHamiltonian` and not into `f1` directly. The intention is that the input here should be a list. Will fix.